ddtrace/internal/writer.py

```diff
@@ -548,4 +558,8 @@ def _stop_service(  # type: ignore[override]
         super(AgentWriter, self)._stop_service()
         self.join(timeout=timeout)
 
-    on_shutdown = periodic
+    def on_shutdown(self):
+        self.periodic()
+        if self._conn:
+            self._conn.close()
+            self._conn = None
```

Comment:
```suggestion
        try:
            self.periodic()
        finally:
            if self._conn:
                self._conn.close()
                self._conn = None
```

Although I assume if periodic shutdown fails we are going down pretty hard so OS will clean up connection?

Reply:
good catch, if periodic does fail it could be possibly that the connection has not been cleaned up.